docs/RFCs/19042021_self_signed_certificates.md

```diff
@@ -0,0 +1,240 @@
+           {{- end }}
+       - name: CA_SECRET
+         values:
+       {{- end }}
+    ```
+ 
+* 3 empty secret will be created in Helm chart for `cockroachdb-ca`, `cockroachdb-node` and `cockroachdb-root` if `tls.certs.generate.enabled`
+ is set.
+  * Data to these secrets will be populated in the `pre-install` job.
+  * In case CA is provided by the user, then `cockroachdb-ca` secret is skipped.
+  * Annotation is set on all the secrets created by CockroachDB; eg: `managed-by: cockroachdb`
+ 
+* A cronjob will be created in Helm chart when `tls.certs.generate.rotateCerts` is set.
+  * This cronjob will run periodically to rotate the certificates.
+  * The schedule of the cronjob will be of two months.
+  * On every scheduled run, it will check if there is any certificate that is going to expire before the next scheduled run,
+   if yes then it will renew the certificates.
+  
+  * <b>The cronjob will use the same `pre-install` job image for certificate rotations. The `pre-install` job image binary will have an argument `--rotate` for handling certificate rotation.</b>
+ 
+* The Stateful pod will be changed to only run `copy-certs` initContainer to copy the certificates from Node secret to emptyDir volume.
+ The rest of the main DB container flow will remain the same.
+ - Client certificate generation process will remain the same as the current implementation using the `post-install` job.
+ 
+`TODO: Need to identify how to generate the SIGHUP signal in all the Nodes for certificate renewal`
+ 
+### Certificate Management Service Implementation
+ 
+ 
+### Overall Flow
+ 
+* Check if CA secret is empty
+  * if yes, [Generate-CA](#generate-ca), [Generate-Node-Cert](#generate-node-cert) and [Generate-Client-Cert](#generate-client-cert); return
+  * if not, [Validate Secret Annotations for CA](#validate-secret-annotations)
+    * if valid, CA is intact
+    * if not, [Annotate-CA](#annotate-ca), [Generate-Node-Cert](#generate-node-cert) and [Generate-Client-Cert](#generate-client-cert); return
+* Check if [CA requires rotation](#check-cert-for-regeneration)
+   * if yes, follow [Rotate-CA](#rotate-ca); return
+* Check if Node or Client certificates [needs to be regenerated](#check-cert-for-regeneration):
+   * if yes, follow [Generate-Node-Cert](#generate-node-cert) and [Generate-Client-Cert](#generate-client-cert)
+ 
+ 
+ 
+#### Generate-CA
+* A self-signed CA will be generated
+* Expiry of the certificate will be driven by the Helm value for `tls.certs.generate.caCertDuration`, passed as env variable
+* Contents of the CA certificate will be stored in the default CA secret `cockroachdb-ca`.
+* An annotation `managed-by: cockroachdb` will be added on secret
+* Follow [CA Annotation Workflow](#annotate-ca)
+ 
+#### Generate-Node-Cert
+* A Node certificate will be generated signed by the generated CA or user custom CA.
+* Contents of the Node certificate will be stored in the Node secret `cockroachdb-node`.
```

Comment:
Need to specify duration of node certificate

Reply:
done